sockeye/inference.py

```diff
@@ -1326,13 +1362,32 @@ def _beam_search(self,
 
             # (7) determine which hypotheses in the beam are now finished
             finished = ((best_word_indices == C.PAD_ID) + (best_word_indices == self.vocab_target[C.EOS_SYMBOL]))
-            if mx.nd.sum(finished).asscalar() == self.batch_size * self.beam_size:  # all finished
-                break
+
+#            self._print_beam(sequences, scores_accumulated, finished, inactive, t)
+
+            if self.beam_search_stop == 'first' and self.batch_size == 1:
+                # TODO: doesn't work for batching
+                if mx.nd.sum(finished).asscalar() > 0:
+                    break
+            else:
+                if mx.nd.sum(finished).asscalar() == self.batch_size * self.beam_size:  # all finished
+                    break
 
             # (8) update models' state with winning hypotheses (ascending)
             for ms in model_states:
                 ms.sort_state(best_hyp_indices)
 
+        # (9) Sort the hypotheses within each sentence (normalization for finished hyps breaks the sort)
+        for sentno in range(self.batch_size):
```

Comment:
this throws us back into the for loop logic before the batched topk change. Can we avoid this?

Reply:
Yes, this can be fixed.